src/leader.h

```diff
@@ -14,61 +14,77 @@
 #define SQLITE_IOERR_LEADERSHIP_LOST (SQLITE_IOERR | (41 << 8))
 
-#define LEADER_NOT_ASYNC INT_MAX
-
 struct exec;
-struct barrier;
 struct leader;
 
-typedef void (*exec_cb)(struct exec *req, int status);
-typedef void (*barrier_cb)(struct barrier *req, int status);
-
-/* Wrapper around raft_apply, saving context information. */
-struct apply {
-	struct raft_apply req; /* Raft apply request */
-	int status;            /* Raft apply result */
-	struct leader *leader; /* Leader connection that triggered the hook */
-	int type;              /* Command type */
-	union {                /* Command-specific data */
-		struct {
-			bool is_commit;
-		} frames;
-	};
-};
+typedef void (*exec_work_cb)(struct exec *req);
+typedef void (*exec_done_cb)(struct exec *req);
+typedef void (*leader_close_cb)(struct leader *l);
 
 struct leader {
-	struct db *db;          /* Database the connection. */
-	sqlite3 *conn;          /* Underlying SQLite connection. */
-	struct raft *raft;      /* Raft instance. */
-	struct exec *exec;      /* Exec request in progress, if any. */
-	queue queue;            /* Prev/next leader, used by struct db. */
-	struct apply *inflight; /* TODO: make leader__close async */
-};
-
-struct barrier {
-	void *data;
-	struct sm sm;
-	struct leader *leader;
-	struct raft_barrier req;
-	barrier_cb cb;
+	void           *data;     /* User data. */
+	struct db      *db;       /* Database for the connection. */
+	sqlite3        *conn;     /* Underlying SQLite connection. */
+	struct raft    *raft;     /* Raft instance. */
+	struct exec    *exec;     /* Exec request in progress, if any. */
+	queue           queue;    /* Prev/next leader, used by struct db. */
+	leader_close_cb close_cb; /* Close callback */
 };
 
 /**
  * Asynchronous request to execute a statement.
  */
 struct exec {
+	/** User data. Not used by the tate machine */
 	void *data;
-	struct sm sm;
-	struct leader *leader;
-	struct barrier barrier;
+
+	/** 
+	 * Already prepared statement to execute. 
+	 * Either this or sql must be set, but not both.
+	 *
+	 * If set, the statement will not be finalized.
+	 * It will always be NULL at the end of the execution.
+	 */
 	sqlite3_stmt *stmt;
+
+	/** 
+	 * SQL statement to execute.
+	 * Either this or stmt must be set, but not both.
+	 *
+	 * The prepared statement will be finalized at the end of the execution.
+	 * It will always be NULL at the end of the execution.
+	 */
+	const char *sql;
+
+	/**
+	 * Tail of the statement. It will be set to the tail of the sql statement
+	 * after a prepare, if set at the start of the execution.
+	 *
+	 * It might be NULL if there is no statement left to execute.
+	 * 
+	 * The lifetime of this pointer is tied to sql field as this will just be
+	 * a pointer to that memory.
+	 */
+	const char *tail;
+
+	/*
+	 * Result code for the operation. RAFT_OK if the operation was successful.
+	 */
 	int status;
+
+	/******* Internal state *******/
```

Comment:
Could you give more context in this comment?

Reply:
Done.